File: cmdServer/protoImpl/httpclient_utils.lua
local M = {}
local utilsFunc = require("buffaloutils/utilsFunc")
local json = require("json")
local enum = require("enum")

local skynet = require("skynet")
local SERVER_NAME = "buffalo/webclient"
--local URL_PREFIX = "https://pay.heywoods.cn/api/"
--local URL_PREFIX = "http://pay_dev.eenpay.cn/api/"
local URL_PREFIX
local APP_ID = enum.APP_ID

local function getUrlPrefix()
	if URL_PREFIX == nil then
		local phpurl_test = skynet.getenv("phpurl_test") or 0
		phpurl_test = tonumber(phpurl_test)
		if phpurl_test == 1 then
			URL_PREFIX = "http://pay_dev.eenpay.cn/api/"
		else
			URL_PREFIX = "https://pay.heywoods.cn/api/"
		end
	end
	return URL_PREFIX
end

local function httppost(player_id,url,postT,bLog)
	local no_reply = false
	local wscAddr = skynet.queryservice(SERVER_NAME)
	local isOk,str = skynet.call(wscAddr,"lua","request",url,nil,postT,no_reply)
	local ret = str
	if type(str) == "string" then
		ret = json.decode(str)
	end
	if bLog then
		utilsFunc.chargePrint(player_id,"url,postT:",url,postT,"isOk,retstr",isOk,str)
		if ret.code ~= 10000 then
			utilsFunc.chargeErrPrint(player_id,"url,postT:",url,postT,"isOk,retstr",isOk,str)
		end
	end
	return isOk,ret
end

local function sign(t)
	local allStrT = {}
	for k,v in pairs(t) do
		allStrT[#allStrT+1] = k .. "=".. v
	end
	table.sort(allStrT)
	return table.concat(allStrT,"&")
end

--------------------------------------------------------------------charge begin
function M.clearBalance(player_id,session_id,amount)
	local urlprefix = getUrlPrefix()
	local url = urlprefix.."orders/midas/create"
	local ct = os.time()
	local randStr = utilsFunc.genRandString(16)
	local getT= {
		appid = APP_ID,
		nonce_str = randStr,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		amount=amount,
		session_id= session_id,
		platform="android",
		goods_info ="clearBalance",
	}

	local signStr = sign(getT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)

	local postT = {
		session_id = session_id,
		amount = amount,
		nonce_str = randStr,
		platform="android",
		goods_info ="clearBalance",

		timestamp = ct,
		appid = APP_ID,
		sign = md5sign,
	}

	local _,creatT= httppost(player_id,url,postT,true)
	local transaction_id = creatT.data.transaction_id

	url = urlprefix.."orders/midas/pay"
	ct = os.time()
	randStr = utilsFunc.genRandString(16)
	postT= {
		appid = APP_ID,
		nonce_str = randStr,
		session_id = session_id,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		transaction_id=tostring(transaction_id),
	}

	signStr = sign( postT)
	md5 = require("md5")
	md5sign = md5.sumhexa(signStr)

	postT.sign = md5sign
	postT.secret = nil

	local isOk,ret= httppost(player_id,url,postT,true)
	if isOk == false then return nil end
	return ret
end


function M.createCharge(player_id,session_id,goods_info,amount,platform)
	local urlprefix = getUrlPrefix()
	local url = urlprefix.."orders/midas/create"
	local ct = os.time()
	local randStr = utilsFunc.genRandString(16)
	local getT= {
		appid = APP_ID,
		nonce_str = randStr,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		goods_info = goods_info,
		amount=amount,
		session_id= session_id,
		platform=platform,
	}

	local signStr = sign(getT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)

	local postT = {
		session_id = session_id,
		goods_info = goods_info,
		amount = amount,
		nonce_str = randStr,
		platform = platform,

		timestamp = ct,
		appid = APP_ID,
		sign = md5sign,
	}

	local isOk,ret = httppost(player_id,url,postT,true)
	if isOk == true then
		return ret
	end
	return nil
end

function M.consumeCharge(player_id,session_id,transaction_id)
	local urlprefix = getUrlPrefix()
	local url = urlprefix.."orders/midas/pay"
	local ct = os.time()
	local randStr = utilsFunc.genRandString(16)
	local postT= {
		appid = APP_ID,
		nonce_str = randStr,
		session_id = session_id,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		transaction_id=transaction_id,
	}

	local signStr = sign( postT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)

	postT.sign = md5sign
	postT.secret = nil

	local isOk,ret = httppost(player_id,url,postT,true)
	if isOk == false then return nil end
	return ret
end

function M.querySum(player_id,session_id,transaction_id)
	local urlprefix = getUrlPrefix()
	local url = urlprefix.."orders/midas/balance"
	local ct = os.time()
	local randStr = utilsFunc.genRandString(16)
	local postT= {
		appid = APP_ID,
		nonce_str = randStr,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		session_id = session_id,
		platform = "android",
	}

	local signStr = sign( postT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)

	postT.sign = md5sign
	postT.secret = nil

	local _,ret =httppost(player_id,url, postT,true)
	return ret
end

function M.queryCharge(player_id,transaction_id)
	local urlprefix = getUrlPrefix()
	local url = urlprefix.."result/query"
	local ct = os.time()
	local randStr = utilsFunc.genRandString(16)
	local postT= {
		appid = APP_ID,
		nonce_str = randStr,
		timestamp = ct,
		secret = enum.MINI_GAME_SECRET,
		transaction_id=transaction_id,
	}

	local signStr = sign( postT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)

	postT.sign = md5sign
	postT.secret = nil

	local _,ret=httppost(player_id,url, postT,true)
	return ret
end
------------------------------------------------------------------charge end
--
local PHP_USR_URL_PREFIX = "https://sdkapi.heywoodsminiprogram.com/user/"
local SALT = "SpPKOMbtVOqh8oLD"
function M.loginCheck(sessionid)
	local url = PHP_USR_URL_PREFIX .. "mini_session_check"
	local ct = os.time()
	local postT = {
		appid = APP_ID,
		sessionid = sessionid,
		timestamp = ct,
		version = "0.1.0",
		sign_salt = SALT,
	}

	local signStr = sign( postT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)
	utilsFunc.debugPrint("loginCheck md5",signStr,md5sign)
	postT.signature = md5sign
	postT.sign_salt = nil

	local _,ret=httppost(0,url, postT,false)
	return ret
end

function M.shareCheck(token)
	local url = PHP_USR_URL_PREFIX .. "mini_share_check"
	local ct = os.time()
	local postT = {
		appid = APP_ID,
		token_type = 1,
		share_token = token,
		timestamp = ct,
		version = "0.1.0",
		sign_salt = SALT,
	}

	local signStr = sign( postT)
	local md5 = require("md5")
	local md5sign = md5.sumhexa(signStr)
	postT.signature = md5sign
	postT.sign_salt = nil

	local _,ret=httppost(0,url, postT,false)
	return ret
end

return M

File: cmdServer/protoImpl/httpclient_utils.go
package protoImpl

import (
	"buffalo/king/common/enumErrCode"
	"buffalo/king/common/gameutil"
	"crypto/md5"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/kingbuffalo/seelog"
	"io/ioutil"
	"net/http"
	"net/url"
	"reflect"
	"strconv"
	"strings"
	"time"
)

const (
	//TODO 这个我不知道是多少
	mini_game_secret string = ""
	platform         string = "android"
)

func getUrlPrefix() string {
	URL_PREFIX := ""
	phpurl_test := enumErrCode.PHPURL_TEST
	if phpurl_test == 1 {
		URL_PREFIX = "http://pay_dev.eenpay.cn/api/"
	} else {
		URL_PREFIX = "https://pay.heywoods.cn/api/"
	}
	return URL_PREFIX
}

func getNow() string {
	now := int(time.Now().Unix())
	cts := strconv.Itoa(now)
	return cts
}
func getMd5(signStr string) string {
	md5sum := md5.Sum([]byte(signStr))
	md5sumStr := fmt.Sprintf("%x", md5sum)
	return md5sumStr
}
func shareCheck(token string) (map[string]interface{}, error) {
	url_ := php_user_url_prefix + "mini_share_check"
	cts := getNow()
	signArr := []string{
		"appid=" + appid,
		"share_token=" + token,
		"sign_salt=" + salt,
		"timestamp=" + cts,
		"token_type=" + "1",
		"version=" + "0.1.0",
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	//发送的一堆数据
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":       {appid},
			"share_token": {token},
			"timestamp":   {cts},
			"version":     {"0.1.0"},
			"token_type":  {"1"},
			"signature":   {md5sumStr},
		})
	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}

func queryCharge(transaction_id string) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "result/query"
	cts := getNow()
	randStr := gameutil.GenRandStr(16)

	signArr := []string{
		"appid=" + appid,
		"nonce_str=" + randStr,
		"secret=" + mini_game_secret,
		"timestamp=" + cts,
		"transaction_id=" + transaction_id,
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	//发送的一堆数据
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":          {appid},
			"nonce_str":      {randStr},
			"timestamp":      {cts},
			"transaction_id": {transaction_id},
			"signature":      {md5sumStr},
		})
	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}

func subClearBalance(transaction_id string, session_id string, cts string) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "orders/midas/pay"
	randStr := gameutil.GenRandStr(16)
	signArr := []string{
		"appid=" + appid,
		"nonce_str=" + randStr,
		"secret=" + mini_game_secret,
		"session_id=" + session_id,
		"timestamp=" + cts,
		"transaction_id=" + transaction_id,
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":          {appid},
			"nonce_str":      {randStr},
			"session_id":     {session_id},
			"timestamp":      {cts},
			"signature":      {md5sumStr},
			"transaction_id": {transaction_id},
		})
	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}
func clearBalance(session_id string, amount int) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "orders/midas/create"
	cts := getNow()
	randStr := gameutil.GenRandStr(16)
	signArr := []string{
		"amount=" + strconv.Itoa(amount),
		"appid=" + appid,
		"goods_info=" + "clearBalance",
		"nonce_str=" + randStr,
		"platform=" + platform,
		"secret=" + mini_game_secret,
		"session_id=" + session_id,
		"timestamp=" + cts,
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	//发送的一堆数据
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":      {appid},
			"amount":     {strconv.Itoa(amount)},
			"goods_info": {"clearBalance"},
			"nonce_str":  {randStr},
			"platform":   {platform},
			"secret":     {mini_game_secret},
			"session_id": {session_id},
			"timestamp":  {cts},
			"signature":  {md5sumStr},
		})
	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	//对jsonM 进行处理, 然后再对
	errCode := jsonM["errCode"]
	var errCodeInt int
	k := reflect.TypeOf(errCode)
	switch k.Kind() {
	case reflect.String:
		var err error
		errCodeInt, err = strconv.Atoi(errCode.(string))
		if err != nil {
			errCodeInt = 1
		}
		break
	case reflect.Float64:
		errCodeInt = int(errCode.(float64))
		break
	}
	if errCodeInt == enumErrCode.ERR_CODE_SUC {
		data := jsonM["data"].(map[string]interface{})
		transaction_id := data["transaction_id"].(string)
		return subClearBalance(transaction_id, session_id, cts)
	}
	errStr := "errCodeInt :" + strconv.Itoa(errCodeInt)
	return nil, errors.New(errStr)
}
func createCharge(session_id string, goods_info string, amount int, platform string) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "orders/midas/create"
	cts := getNow()
	randStr := gameutil.GenRandStr(16)
	signArr := []string{
		"amount=" + strconv.Itoa(amount),
		"appid=" + appid,
		"goods_info=" + goods_info,
		"nonce_str=" + randStr,
		"platform=" + platform,
		"secret=" + mini_game_secret,
		"session_id=" + session_id,
		"timestamp=" + cts,
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":      {appid},
			"amount":     {strconv.Itoa(amount)},
			"nonce_str":  {randStr},
			"goods_info": {goods_info},
			"platform":   {platform},
			"timestamp":  {cts},
			"session_id": {session_id},
			"signature":  {md5sumStr},
		})

	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}

func consumeCharge(session_id string, transaction_id string) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "orders/midas/pay"
	cts := getNow()
	randStr := gameutil.GenRandStr(16)
	signArr := []string{
		"appid=" + appid,
		"nonce_str=" + randStr,
		"secret=" + mini_game_secret,
		"session_id=" + session_id,
		"timestamp=" + cts,
		"transaction_id=" + transaction_id,
	}

	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":          {appid},
			"nonce_str":      {randStr},
			"timestamp":      {cts},
			"session_id":     {session_id},
			"transaction_id": {transaction_id},
			"signature":      {md5sumStr},
		})

	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}
func querySum(session_id string, transaction_id string) (map[string]interface{}, error) {
	url_ := getUrlPrefix() + "orders/midas/balance"
	cts := getNow()
	randStr := gameutil.GenRandStr(16)

	signArr := []string{
		"appid=" + appid,
		"nonce_str=" + randStr,
		"platform=" + platform,
		"secret=" + mini_game_secret,
		"session_id=" + session_id,
		"timestamp=" + cts,
	}
	signStr := strings.Join(signArr, "&")
	md5sumStr := getMd5(signStr)
	//发送的一堆数据
	resp, err := http.PostForm(url_,
		url.Values{
			"appid":      {appid},
			"nonce_str":  {randStr},
			"timestamp":  {cts},
			"platform":   {platform},
			"session_id": {session_id},
			"signature":  {md5sumStr},
		})

	if err != nil {
		return nil, err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			seelog.Debug("respBodyCloseErr", err)
		}
	}()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	seelog.Trace("body", string(body))
	var jsonM map[string]interface{}
	err = json.Unmarshal(body, &jsonM)
	if err != nil {
		return nil, err
	}
	return jsonM, nil
}
